### DISCUSSION TITLE
multi UEs+ZeroMQ+multi gnbs +open5gs

### DISCUSSION THREAD

**Discussion by wmwwimor** [srsran/srsRAN_4G#1203] [Q&A] [answered]
I would like to know if it is possible to implement this architecture:multi UEs+ZeroMQ+one gnb +open5gs. Can it currently be achieved using GNU Radio? Is it possible to further expand it into such an architecture:multi UEs+ZeroMQ+multi gnbs +open5gs? I don't want to use physical devices. Is there something similar to ZeroMQ available now that can assist in simulating radio frequencies?

**Comment by brendan-mcauliffe** [comment, marked as answer]
Hi @wmwwimor, 

Yes, this architecture is possible. You can use ZMQ and GNU Radio. Here are a couple of pointers: 

- Each UE will need its own config file, ID, and namespace so that you can correctly send messages to and from the UE. 
- You can expand this to add multiple gNBs, you need to make sure the gNBs use different IDs. 

If you plan to do 5G testing with srsUE and ZMQ, I suggest you use the srsRAN Project gNB instead of the srsRAN 4G gNB. You can find the new gNB code base [here](https://github.com/srsran/srsRAN_project). 

See the following screenshot of a GNU Radio flowgraph that supports a multi-UE scenario with a single gNB: 
![image](https://github.com/srsran/srsRAN_4G/assets/37801273/b879dc52-d696-497b-ba5d-8dac36676702)

**Reply by arrowtem** [reply]
is it possible to launch several ue on 1 b200? using add function from gnu radio

**Reply by nishaadh** [reply]
Could you expand a bit more on adding multiple gNBs? Do they use the same Open5gs core?

**Comment by wmwwimor** [comment]
I tried to attach 2UEs via Gnuradio at the same time ,I got the following issue in open5gs log :
07/30 13:47:42.372: [amf] INFO: [suci-0-001-01-0000-0-0-0123456780] Unknown UE by SUCI (../src/amf/context.c:1776)
07/30 13:47:42.372: [amf] INFO: [Added] Number of AMF-UEs is now 1 (../src/amf/context.c:1563)
07/30 13:47:42.372: [gmm] INFO: Registration request (../src/amf/gmm-sm.c:985)
07/30 13:47:42.372: [gmm] INFO: [suci-0-001-01-0000-0-0-0123456780]    SUCI (../src/amf/gmm-handler.c:152)
07/30 13:48:12.395: [gmm] WARNING: [suci-0-001-01-0000-0-0-0123456780] Retransmission failed. Stop retransmission (../src/amf/gmm-sm.c:1440)
07/30 13:48:12.396: [amf] WARNING: [suci-0-001-01-0000-0-0-0123456780] Authentication reject (../src/amf/nas-path.c:538)
07/30 13:48:12.396: [amf] INFO: UE Context Release [Action:3] (../src/amf/ngap-handler.c:1632)
07/30 13:48:12.396: [amf] INFO:     RAN_UE_NGAP_ID[0] AMF_UE_NGAP_ID[1] (../src/amf/ngap-handler.c:1633)
07/30 13:48:12.396: [amf] INFO:     SUCI[suci-0-001-01-0000-0-0-0123456780] (../src/amf/ngap-handler.c:1636)
07/30 13:48:12.396: [amf] INFO: [Removed] Number of gNB-UEs is now 0 (../src/amf/context.c:2517)
07/30 13:48:12.396: [amf] INFO: [Removed] Number of AMF-UEs is now 0 (../src/amf/context.c:1656)

Both ues have RRC connection , but neither of them got an IP address.

**Reply by small-hulu** [reply]
Hello, I am also having the same problem, is the problem solved?how was it solved?thank you。

**Reply by raul-parada** [reply]
Hi. Did someone make it work? I have built the above flowchart on GNU Radio. I have some doubts: 1) Do we need to create the ue.conf from all UEs? I have created 3 additional UEs with the following arguments: `device_args = tx_port=tcp://127.0.0.1:2X00,rx_port=tcp://127.0.0.1:2X01,base_srate=11.52e6` is it correct? If run the ue.conf then the address is already used and by running the flowchart I receive that the bind address is already used. If I execute the processes vice versa, I receive the following error when executing the ue.conf ```
Error: binding transmitter socket (tcp://127.0.0.1:2001): Address already in use
[zmq] Error: opening transmitter
Error initializing radio.

Please, could you explain the steps you've followed? BTW, I'm using this tutorial deployment ```https://docs.srsran.com/projects/project/en/latest/tutorials/source/flexric/source/index.html#gnb

**Reply by connor01M** [reply]
Was someone able to get it to work with multiple GNBs? I haven't found a way to setup the config files for that to work correctly

**Reply by pgawlowicz** [reply]
@wmwwimor  @connor01M Please check the new section of our gnb+srsUE tutorial that describes how to emulate multiple-UEs:
https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#multi-ue-emulation

You can use it as a starting point for your multi-gnb/multi-ue scenario. 

Please remember to clone the newest version of the srsRAN-4G repository.

Please provide your feedback.
